### PR TITLE
add a failure reason for the operator test cases

### PR DESCRIPTION
**Pull request by aabughosh**

### CODE REVIEW

File: cnf-certification-test/operator/suite.go [hunk 1]
```diff
@@ -104,23 +108,28 @@ func testOperatorInstallationWithoutPrivileges(env *provider.TestEnvironment) {
...
-	nonCompliantCsvs := []string{}
+	var compliantObjects []*testhelper.ReportObject
+	var nonCompliantObjects []*testhelper.ReportObject
 
 	for i := range env.Operators {
 		operator := env.Operators[i]
 		if operator.SubscriptionName == "" {
 			tnf.ClaimFilePrintf("OLM subscription not found for operator from csv %s", provider.CsvToString(operator.Csv))
-			nonCompliantCsvs = append(nonCompliantCsvs, provider.CsvToString(operator.Csv))
+			nonCompliantObjects = append(nonCompliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "OLM subscription not found for operator, so its not installed via OLM", false))
```
Comment (jmontesi) [inline cnf-certification-test/operator/suite.go:128]:
```suggestion
			nonCompliantObjects = append(nonCompliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "OLM subscription not found for operator, so it is not installed via OLM", false))
```

**Review by jmontesi — APPROVED** [review]

**Review by sebrandon1 — APPROVED** [review]
One string suggestion

File: cnf-certification-test/operator/suite.go [hunk 2]
```diff
@@ -70,17 +71,20 @@ func testOperatorInstallationPhaseSucceeded(env *provider.TestEnvironment) {
 		// could be restarting. Let's give it some time before declaring it failed.
 		phase := phasecheck.WaitOperatorReady(csv)
 		if phase != v1alpha1.CSVPhaseSucceeded {
-			badOperators = append(badOperators, env.Operators[i].String())
+			nonCompliantObjects = append(nonCompliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "Operator not in Succeeded state ", false))
```
Comment (edcdavid) [inline cnf-certification-test/operator/suite.go:74]: Add a new field for the operator named phase or operatorPhase and add it to the operator object with addField

File: cnf-certification-test/operator/suite.go [hunk 3]
```diff
@@ -70,17 +71,20 @@ func testOperatorInstallationPhaseSucceeded(env *provider.TestEnvironment) {
...
 			tnf.ClaimFilePrintf("%s is in phase %s. Expected phase is %s",
 				&env.Operators[i], csv.Status.Phase, v1alpha1.CSVPhaseSucceeded)
+		} else {
+			compliantObjects = append(compliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "Operator on Succeeded state ", true))
```
Comment (edcdavid) [inline cnf-certification-test/operator/suite.go:78]: same here

File: cnf-certification-test/operator/suite.go [hunk 4]
```diff
@@ -104,23 +108,28 @@ func testOperatorInstallationWithoutPrivileges(env *provider.TestEnvironment) {
 		}
```
Comment (edcdavid) [inline cnf-certification-test/operator/suite.go:108]: add the compliant/non compliant object in the loop and if :
if n := len(permission.Rules[ruleIndex].ResourceNames); n > 0 
and add a rules and permission field to the operator

File: cnf-certification-test/operator/suite.go [hunk 5]
```diff
@@ -104,23 +108,28 @@ func testOperatorInstallationWithoutPrivileges(env *provider.TestEnvironment) {
...
 
 		if badRuleFound {
-			badOperators = append(badOperators, env.Operators[i].String())
+			nonCompliantObjects = append(nonCompliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "Operator has privileges on cluster resources ", false))
```
Comment (edcdavid) [inline cnf-certification-test/operator/suite.go:111]: Also keep this level (whole operator reporting level)

File: cnf-certification-test/operator/suite.go [hunk 6]
```diff
@@ -104,23 +108,28 @@ func testOperatorInstallationWithoutPrivileges(env *provider.TestEnvironment) {
...
-	nonCompliantCsvs := []string{}
+	var compliantObjects []*testhelper.ReportObject
+	var nonCompliantObjects []*testhelper.ReportObject
 
 	for i := range env.Operators {
 		operator := env.Operators[i]
 		if operator.SubscriptionName == "" {
 			tnf.ClaimFilePrintf("OLM subscription not found for operator from csv %s", provider.CsvToString(operator.Csv))
-			nonCompliantCsvs = append(nonCompliantCsvs, provider.CsvToString(operator.Csv))
+			nonCompliantObjects = append(nonCompliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "OLM subscription not found for operator, so it is not installed via OLM", false))
```
Comment (edcdavid) [inline cnf-certification-test/operator/suite.go:128]: Add a subscription name field to the operator reporting object and use the addField function to add it

File: cnf-certification-test/operator/suite.go [hunk 7]
```diff
@@ -99,28 +107,40 @@ func testOperatorInstallationWithoutPrivileges(env *provider.TestEnvironment) {
 						env.Operators[i], permission.ServiceAccountName, n, ruleIndex)
 					// Keep reviewing other permissions' rules so we can log all the failing ones in the claim file.
 					badRuleFound = true
+					nonCompliantObjects = append(nonCompliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "Operator has privileges on cluster resources ", false).
+						AddField(testhelper.ServiceAccountName, permission.ServiceAccountName).AddField(testhelper.ResourceName, strings.Join(permission.Rules[ruleIndex].ResourceNames, "")))
+				} else {
+					compliantObjects = append(compliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "Operator has no privileges on cluster resources", true).
+						AddField(testhelper.ServiceAccountName, permission.ServiceAccountName).AddField(testhelper.ResourceName, strings.Join(permission.Rules[ruleIndex].ResourceNames, "")))
 				}
```
Comment (edcdavid) [inline cnf-certification-test/operator/suite.go:115]: no need to add the ResourceName field here as it is always empty

File: cnf-certification-test/operator/suite.go [hunk 8]
```diff
@@ -99,28 +107,40 @@ func testOperatorInstallationWithoutPrivileges(env *provider.TestEnvironment) {
 						env.Operators[i], permission.ServiceAccountName, n, ruleIndex)
 					// Keep reviewing other permissions' rules so we can log all the failing ones in the claim file.
 					badRuleFound = true
+					nonCompliantObjects = append(nonCompliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "Operator has privileges on cluster resources ", false).
+						AddField(testhelper.ServiceAccountName, permission.ServiceAccountName).AddField(testhelper.ResourceName, strings.Join(permission.Rules[ruleIndex].ResourceNames, "")))
+				} else {
+					compliantObjects = append(compliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "Operator has no privileges on cluster resources", true).
+						AddField(testhelper.ServiceAccountName, permission.ServiceAccountName).AddField(testhelper.ResourceName, strings.Join(permission.Rules[ruleIndex].ResourceNames, "")))
```
Comment (edcdavid) [inline cnf-certification-test/operator/suite.go:114]: strings.Join(permission.Rules[ruleIndex].ResourceNames, this is always empty. Replace with "n/a" or similar

add "s" for plural to testhelper.ResourceName -> testhelper.ResourceName+"s"

**Comment by sebrandon1** [comment]
@aabughosh can you also rebase this PR so you pick up the changes in #1306?  Thanks!

File: cnf-certification-test/operator/suite.go [hunk 9]
```diff
@@ -99,28 +107,40 @@ func testOperatorInstallationWithoutPrivileges(env *provider.TestEnvironment) {
 						env.Operators[i], permission.ServiceAccountName, n, ruleIndex)
 					// Keep reviewing other permissions' rules so we can log all the failing ones in the claim file.
 					badRuleFound = true
+					nonCompliantObjects = append(nonCompliantObjects, testhelper.NewOperatorReportObject(env.Operators[i].Namespace, env.Operators[i].Name, "Operator has privileges on cluster resources ", false).
```
[the remark this answers is not in the document]
Comment (aabughosh) [reply cnf-certification-test/operator/suite.go:110]: there is no operator type u mean for the ServiceAccountName?

**Review by edcdavid — APPROVED** [review]

**Comment by sebrandon1** [comment]
```
FATAL  [Aug  8 14:09:29.660][suite.go: 102] failed running preflight on image: quay.io/testnetworkfunction/cnf-test-partner:latest error: failed to extract tarball: write /tmp/preflight-795648300/fs/usr/share/doc/systemd/NEWS: no space left on device 
```

This was fixed in #1320.

**Comment by sebrandon1** [comment]
We'll merge when the `lifecycle` QE tests pass.

**Pull request merged**